core/src/main/scala/cats/Eval.scala

```diff
@@ -134,23 +136,57 @@ final case class Now[A](value: A) extends Eval[A] {
  * by the closure) will not be retained, and will be available for
  * garbage collection.
  */
-final class Later[A](f: () => A) extends Eval[A] {
-  private[this] var thunk: () => A = f
-
-  // The idea here is that `f` may have captured very large
-  // structures, but produce a very small result. In this case, once
-  // we've calculated a value, we would prefer to be able to free
-  // everything else.
-  //
-  // (For situations where `f` is small, but the output will be very
-  // expensive to store, consider using `Always`.)
-  lazy val value: A = {
-    val result = thunk()
-    thunk = null // scalastyle:off
-    result
-  }
+final class Later[A](f: () => A) extends Eval[A] with Externalizable {
+
+  // we use lock to prevent possible problems with someone else
+  // synchronizing on 'this'.
+  private[this] val lock = new Object()
+  @volatile private[this] var thunk: Function0[A] = f
+  private[this] var result: A = _
+
+  final def value: A =
+    if (thunk == null) {
+      result
```

Comment:
Yea, agree - switching the `volatile` to `result` should close any loops. I don't think you need a cast though - at least not an explicit one - this should be fine.

```scala
 final class Later[A](f: () => A) extends Eval[A] {

   // we use lock to prevent possible problems with someone else
   // synchronizing on 'this'.
   private[this] val lock = new Object()
   private[this] var thunk: Function0[A] = f

   @volatile private[this] var result: A = _

   final def value: A =
     if (result != null) {
       result
     } else lock.synchronized {
       if (result != null) result else {
         result = thunk()
         thunk = null
         result
       }
     }

   def memoize: Eval[A] = this
 }
```
BTW - here is an alternate encoding. Not sure what would be better - but it is an interesting problem so I was thinking of a few different approaches and this one seemed interesting.

```scala
 final class Later[A](f: () => A) extends Eval[A] {

   private[this] var thunk: () => A = f

   private[this] trait Result {
     def value: A
   }

   private[this] val evaluated: Result = new Result {
     lazy val value = {
       val result = thunk()
       thunk = null
       result
     }
   }

   final def value: A = evaluated.value

   def memoize: Eval[A] = this
 }
```
This encoding avoids explicit use of `volatile` or `synchronized`, and moves the lock to an inner type so should avoid the `synchronization` problem mentioned. I'm not sure how the cost of the additional reference to the `lazy val `in the inner type would affect performance, I guess it would need to be measured.